File: deprecated/ClassicLib/messaging/core/enums.py
"""Message type and target enums for the MessageHandler system.

This module provides enumerations for message categorization and routing.
These interoperate directly with Rust enums via the `classic_message` module.

"""

from __future__ import annotations

from enum import Enum, auto
from typing import TYPE_CHECKING, Any

from classic_message import MessageTarget as RustMessageTarget
from classic_message import MessageType as RustMessageType

if TYPE_CHECKING:
    from typing import Self


class MessageType(Enum):
    """Message type enumeration for categorizing messages.

    Attributes:
        INFO: Informational message.
        WARNING: Warning message indicating a potential issue.
        ERROR: Error message indicating a failure or problem.
        SUCCESS: Message indicating successful completion.
        PROGRESS: Message signaling progress of an ongoing operation.
        DEBUG: Message intended for debugging purposes.
        CRITICAL: Message indicating a critical issue requiring attention.

    """

    INFO = auto()
    WARNING = auto()
    ERROR = auto()
    SUCCESS = auto()
    PROGRESS = auto()
    DEBUG = auto()
    CRITICAL = auto()

    def to_rust(self) -> Any:
        """Convert to Rust MessageType.

        Returns:
            Rust MessageType enum value.

        """
        mapping = {
            MessageType.INFO: RustMessageType.Info,
            MessageType.WARNING: RustMessageType.Warning,
            MessageType.ERROR: RustMessageType.Error,
            MessageType.SUCCESS: RustMessageType.Success,
            MessageType.PROGRESS: RustMessageType.Progress,
            MessageType.DEBUG: RustMessageType.Debug,
            MessageType.CRITICAL: RustMessageType.Critical,
        }
        return mapping[self]

    @classmethod
    def from_rust(cls, rust_type: Any) -> Self:
        """Convert from Rust MessageType.

        Args:
            rust_type: Rust MessageType enum value.

        Returns:
            Corresponding Python MessageType.

        Raises:
            ValueError: If unknown Rust type.

        """
        # Use the name for mapping
        name = str(rust_type).lower()
        for member in cls:
            if member.name.lower() == name:
                return member

        msg = f"Unknown Rust MessageType: {rust_type}"
        raise ValueError(msg)


class MessageTarget(Enum):
    """Message target enumeration for routing messages.

    Attributes:
        ALL: Show in both GUI and CLI.
        GUI: Show only in GUI mode.
        CONSOLE: Show only in CLI mode.
        LOG_ONLY: Only write to log file, no display.

    """

    ALL = auto()
    GUI = auto()
    CONSOLE = auto()
    LOG_ONLY = auto()

    def normalize(self) -> MessageTarget:
        """Return self (no normalization needed with canonical values only).

        Returns:
            The same MessageTarget value.

        """
        return self

    def should_display_in_gui(self) -> bool:
        """Check if message should display in GUI mode.

        Returns:
            True if should display in GUI.

        """
        normalized = self.normalize()
        return normalized in {MessageTarget.ALL, MessageTarget.GUI}

    def should_display_in_cli(self) -> bool:
        """Check if message should display in CLI mode.

        Returns:
            True if should display in CLI.

        """
        normalized = self.normalize()
        return normalized in {MessageTarget.ALL, MessageTarget.CONSOLE}

    def should_display(self) -> bool:
        """Check if message should display at all (not log-only).

        Returns:
            True if should display.

        """
        normalized = self.normalize()
        return normalized != MessageTarget.LOG_ONLY

    def to_rust(self) -> Any:
        """Convert to Rust MessageTarget.

        Returns:
            Rust MessageTarget enum value.

        """
        # Normalize first, then map
        normalized = self.normalize()
        mapping = {
            MessageTarget.ALL: RustMessageTarget.All,
            MessageTarget.GUI: RustMessageTarget.Gui,
            MessageTarget.CONSOLE: RustMessageTarget.Console,
            MessageTarget.LOG_ONLY: RustMessageTarget.LogOnly,
        }
        return mapping[normalized]

    @classmethod
    def from_rust(cls, rust_target: Any) -> MessageTarget:
        """Convert from Rust MessageTarget.

        Args:
            rust_target: Rust MessageTarget enum value.

        Returns:
            Corresponding Python MessageTarget.

        Raises:
            ValueError: If unknown Rust target.

        """
        # Use string representation for mapping
        name = str(rust_target).lower()
        # Map to canonical names
        if "gui" in name:
            return cls.GUI
        if "console" in name or "cli" in name:
            return cls.CONSOLE
        if "all" in name:
            return cls.ALL
        if "log" in name:
            return cls.LOG_ONLY

        msg = f"Unknown Rust MessageTarget: {rust_target}"
        raise ValueError(msg)

File: business-logic/classic-message-core/src/enums.rs
//! Message type and target enumerations for the MessageHandler system.
use serde::{Deserialize, Serialize};

/// Represents various types of messages categorized by their context and severity.
///
/// This enumeration helps to define and manage message categories efficiently
/// for logging or displaying purposes. It provides a predefined set of message
/// types that can be used across an application to standardize how messages are
/// identified and processed.
///
/// # Examples
///
/// ```rust
/// use classic_message_core::MessageType;
///
/// let msg_type = MessageType::Info;
/// assert_eq!(msg_type.to_log_level(), log::Level::Info);
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum MessageType {
    /// Denotes an informational message.
    Info,
    /// Denotes a warning message indicating a potential issue.
    Warning,
    /// Denotes an error message indicating a failure or problem.
    Error,
    /// Denotes a message indicating the successful completion of an operation or process.
    Success,
    /// Denotes a message signaling the progress of an ongoing operation.
    Progress,
    /// Denotes a message intended for debugging purposes.
    Debug,
    /// Denotes a message indicating a critical issue requiring immediate attention.
    Critical,
}

impl MessageType {
    /// Converts the message type to its corresponding log level.
    ///
    /// Maps message types to standard logging levels for consistent logging behavior:
    /// - Info, Success → log::Level::Info
    /// - Warning → log::Level::Warn
    /// - Error, Critical → log::Level::Error
    /// - Debug, Progress → log::Level::Debug
    ///
    /// # Returns
    ///
    /// The corresponding `log::Level` for this message type.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageType;
    ///
    /// assert_eq!(MessageType::Error.to_log_level(), log::Level::Error);
    /// assert_eq!(MessageType::Info.to_log_level(), log::Level::Info);
    /// ```
    #[must_use]
    pub const fn to_log_level(&self) -> log::Level {
        match self {
            Self::Debug | Self::Progress => log::Level::Debug,
            Self::Info | Self::Success => log::Level::Info,
            Self::Warning => log::Level::Warn,
            Self::Error | Self::Critical => log::Level::Error,
        }
    }

    /// Returns a human-readable name for the message type.
    ///
    /// # Returns
    ///
    /// A string slice representing the name of the message type.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageType;
    ///
    /// assert_eq!(MessageType::Info.name(), "Info");
    /// assert_eq!(MessageType::Critical.name(), "Critical");
    /// ```
    #[must_use]
    pub const fn name(&self) -> &'static str {
        match self {
            Self::Info => "Info",
            Self::Warning => "Warning",
            Self::Error => "Error",
            Self::Success => "Success",
            Self::Progress => "Progress",
            Self::Debug => "Debug",
            Self::Critical => "Critical",
        }
    }
}

/// Enumeration to specify where a message should be directed.
///
/// This enum defines various message targets such as both GUI and CLI,
/// only GUI, only CLI, or logs. It is used to control and direct how
/// messages are processed and displayed in an application context.
///
/// # Examples
///
/// ```rust
/// use classic_message_core::MessageTarget;
///
/// let target = MessageTarget::All;
/// assert!(target.should_display_in_gui());
/// assert!(target.should_display_in_cli());
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
pub enum MessageTarget {
    /// Show in both GUI and CLI.
    #[default]
    All,
    /// Show only in GUI mode (legacy, prefer `Gui`).
    GuiOnly,
    /// Show only in CLI mode (legacy, prefer `Console`).
    CliOnly,
    /// Only write to log file, no display.
    LogOnly,
    /// Show only in GUI mode.
    Gui,
    /// Show only in CLI mode.
    Console,
}

impl MessageTarget {
    /// Determines if the message should be displayed in GUI mode.
    ///
    /// # Arguments
    ///
    /// * `is_gui_mode` - Whether the application is currently in GUI mode.
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed in GUI, `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display_in_gui());
    /// assert!(MessageTarget::Gui.should_display_in_gui());
    /// assert!(!MessageTarget::Console.should_display_in_gui());
    /// assert!(!MessageTarget::LogOnly.should_display_in_gui());
    /// ```
    #[must_use]
    pub const fn should_display_in_gui(&self) -> bool {
        matches!(self, Self::All | Self::GuiOnly | Self::Gui)
    }

    /// Determines if the message should be displayed in CLI mode.
    ///
    /// # Arguments
    ///
    /// * `is_gui_mode` - Whether the application is currently in GUI mode.
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed in CLI, `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display_in_cli());
    /// assert!(MessageTarget::Console.should_display_in_cli());
    /// assert!(!MessageTarget::Gui.should_display_in_cli());
    /// assert!(!MessageTarget::LogOnly.should_display_in_cli());
    /// ```
    #[must_use]
    pub const fn should_display_in_cli(&self) -> bool {
        matches!(self, Self::All | Self::CliOnly | Self::Console)
    }

    /// Determines if the message should be displayed at all (not log-only).
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed (not log-only), `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display());
    /// assert!(!MessageTarget::LogOnly.should_display());
    /// ```
    #[must_use]
    pub const fn should_display(&self) -> bool {
        !matches!(self, Self::LogOnly)
    }
}

#[cfg(test)]
#[path = "enums_tests.rs"]
mod tests;
